File: scripts/backfill-entities.ts
/**
 * One-time backfill: iterates all existing bills in bill_embeddings,
 * fetches their XML from govinfo.gov, and populates the entity columns.
 *
 * Run: npx tsx scripts/backfill-entities.ts
 *
 * Resume-safe: skips bills where referenced_agencies is already non-empty.
 * Set FORCE=1 to reprocess all bills regardless.
 */

import { createServiceClient } from '@/lib/supabase/service'
import { fetchBillTextXml, extractTextFromBillXml } from './lib/fetch-bill-text'
import { extractAgencies } from './lib/federal-agencies'
import { extractCitations } from './lib/parse-citations'

const DELAY_MS = 200   // Between govinfo requests — be respectful
const FORCE = process.env.FORCE === '1'

async function sleep(ms: number) {
  return new Promise(r => setTimeout(r, ms))
}

async function backfillEntities() {
  const supabase = createServiceClient()

  let query = supabase
    .from('bill_embeddings')
    .select('bill_id, congress')
    .order('introduced_date', { ascending: false })

  if (!FORCE) {
    // Only process bills not yet populated
    query = query.eq('referenced_agencies', '{}')
  }

  const { data: bills, error } = await query

  if (error) {
    console.error('Failed to fetch bills:', error.message)
    process.exit(1)
  }

  const total = bills?.length ?? 0
  console.log(`Backfilling ${total} bills (FORCE=${FORCE})…`)

  let processed = 0
  let updated = 0
  let skipped = 0

  for (const bill of bills ?? []) {
    const parts = bill.bill_id.split('-')
    if (parts.length < 3) { skipped++; continue }

    const congress = parseInt(parts[0], 10)
    const type = parts[1]
    const number = parts.slice(2).join('-') // handles multi-part numbers

    const xml = await fetchBillTextXml(congress, type, number)

    if (xml) {
      const fullText = extractTextFromBillXml(xml)
      const agencies = extractAgencies(fullText)
      const citations = extractCitations(fullText)

      const { error: updateError } = await supabase
        .from('bill_embeddings')
        .update({
          referenced_agencies: agencies,
          referenced_laws: [...citations.actNames, ...citations.publicLaws],
          referenced_usc: citations.uscSections,
        })
        .eq('bill_id', bill.bill_id)

      if (updateError) {
        console.error(`  [${bill.bill_id}] update error:`, updateError.message)
      } else {
        updated++
      }
    } else {
      skipped++
    }

    processed++
    if (processed % 25 === 0 || processed === total) {
      console.log(`  ${processed}/${total} processed — ${updated} updated, ${skipped} skipped`)
    }

    await sleep(DELAY_MS)
  }

  console.log(`\nDone. ${updated} bills updated, ${skipped} skipped (no XML or parse error).`)
}

backfillEntities().catch(err => {
  console.error(err)
  process.exit(1)
})

File: pipeline/scripts/backfill_entities.py
"""
One-time backfill: iterates all existing bills in bill_embeddings,
fetches their XML from govinfo.gov, and populates the entity columns.

Run: python pipeline/scripts/backfill_entities.py

Resume-safe: skips bills where referenced_agencies is already non-empty.
Set FORCE=1 to reprocess all bills regardless.
"""
import os
import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent.parent))

from pipeline.lib.supabase_client import create_service_client
from pipeline.lib.fetch_bill_text import fetch_bill_text_xml, extract_text_from_bill_xml
from pipeline.lib.federal_agencies import extract_agencies
from pipeline.lib.parse_citations import extract_citations
from pipeline.lib.config import RATE_LIMIT_GOVINFO

FORCE = os.environ.get("FORCE") == "1"


def backfill_entities() -> None:
    supabase = create_service_client()

    query = (
        supabase.table("bills")
        .select("bill_id, congress")
        .order("introduced_date", desc=True)
    )
    if not FORCE:
        query = query.eq("referenced_agencies", "{}")

    resp = query.execute()
    bills = resp.data or []

    total = len(bills)
    print(f"Backfilling {total} bills (FORCE={FORCE})...")

    processed = updated = skipped = 0

    for bill in bills:
        parts = bill["bill_id"].split("-")
        if len(parts) < 3:
            skipped += 1
            continue

        congress = int(parts[0])
        bill_type = parts[1]
        number = "-".join(parts[2:])

        xml = fetch_bill_text_xml(congress, bill_type, number)

        if xml:
            full_text = extract_text_from_bill_xml(xml)
            agencies = extract_agencies(full_text)
            citations = extract_citations(full_text)

            update_resp = (
                supabase.table("bills")
                .update({
                    "referenced_agencies": agencies,
                    "referenced_laws":     citations.act_names + citations.public_laws,
                    "referenced_usc":      citations.usc_sections,
                })
                .eq("bill_id", bill["bill_id"])
                .execute()
            )
            updated += 1
        else:
            skipped += 1

        processed += 1
        if processed % 25 == 0 or processed == total:
            print(f"  {processed}/{total} processed — {updated} updated, {skipped} skipped")

        time.sleep(RATE_LIMIT_GOVINFO)

    print(f"\nDone. {updated} bills updated, {skipped} skipped (no XML or parse error).")


if __name__ == "__main__":
    backfill_entities()
